# Data Preprocessing

In [1]:
import pandas as pd
import numpy as np

## Membaca dan menggabungkan semua sheet

Membaca data tapi karena di file data ada beberapa sheet harus digabungkan terlebih dahulu

In [2]:
dfs = pd.read_excel(
    '../dataset/ML engineer jobs (aug 30, 2025).xlsx',
    sheet_name=None,
    header=[0,1]
)

print(len(dfs.values()))

df_list = []
for sheet_name, d in dfs.items():
    d["Platform"] = sheet_name   # kolom baru sesuai nama sheet
    df_list.append(d)

# concat list, bukan dfs.values()
df = pd.concat(df_list, ignore_index=True)

# hapus kolom "No"

df.head()

6


(No, Unnamed: 0_level_1) (Company, Unnamed: 1_level_1)  \
0                       1.0                   Braintrust    
1                       2.0         VIDA Digital Identity   
2                       3.0                   RH USA, Inc   
3                       4.0                      Bibit.id   
4                       5.0        CAD IT Gruppo Cedacri    

                      (Position, Unnamed: 2_level_1)  \
0                      Prompt Engineer / AI Engineer   
1  Deep Learning / AI Scientist - Liveness Detect...   
2                               AI Prompter Engineer   
3                                     AI/ML Engineer   
4                             Computer Vision Intern   

         (Location , Unnamed: 3_level_1)  \
0                Asia Pasifik dan Jepang   
1       Jakarta, Jakarta Raya, Indonesia   
2  Bekasi Selatan, Jawa Barat, Indonesia   
3      Jakarta, Jakarta Raya, Indonesia    
4        Gedebage, Jawa Barat, Indonesia   

  (WFH/ WFO/ Hybird, Unnamed: 4_level_1)  \
0                                    WFH   
1                                      -   
2                                    WFO   
3                                 Hybird   
4                                    WFO   

  (length of work experience required (years), MIN)  \
0                                                 3   
1                                                 3   
2                                                 0   
3                                                 0   
4                                                 0   

  (length of work experience required (years), MAX)  \
0                                                 5   
1                                                 -   
2                                                 -   
3                                                 -   
4                                                 -   

  (Salary, Unnamed: 7_level_1) (Access Date, Unnamed: 8_level_1)  \
0                            -                        2025-08-30   
1                            -                        2025-08-30   
2                            -                        2025-08-30   
3                            -                        2025-08-30   
4                 RP 1,5 - 2 M                        2025-08-30   

  (Upload By Company, Unnamed: 9_level_1)  ... (Degree, Unnamed: 12_level_1)  \
0                               - 2 weeks  ...                             -   
1                  30-Agu-2025 (reupload)  ...               master’s or PhD   
2                                - 6 days  ...                    Bachelor's   
3                               - 1 Month  ...        Bachelor's or Master's   
4                     2025-08-30 00:00:00  ...                       Diploma   

  (Information, Unnamed: 13_level_1) (Platform, )  \
0  Respons dikelola di luar LinkedIn     Linkedin   
1  Respons dikelola di luar LinkedIn     Linkedin   
2  Respons dikelola di luar LinkedIn     Linkedin   
3                                  -     Linkedin   
4  Respons dikelola di luar LinkedIn     Linkedin   

  (Category, Unnamed: 10_level_1) (Degree, Unnamed: 11_level_1) Platform  \
0                             NaN                           NaN      NaN   
1                             NaN                           NaN      NaN   
2                             NaN                           NaN      NaN   
3                             NaN                           NaN      NaN   
4                             NaN                           NaN      NaN   

  (Quota, Unnamed: 11_level_1) (Category, Unnamed: 12_level_1)  \
0                          NaN                             NaN   
1                          NaN                             NaN   
2                          NaN                             NaN   
3                          NaN                             NaN   
4                          NaN                             NaN   

   (Degree, Unnamed: 13_level_1) (Informatio

Merapihkan nama kolom dengan menggabungkan yang unnamed pada level 0 dan 1 dan setelahnya menghilangkan kolom No

In [3]:
df.columns = [
    col[0] if "Unnamed" in str(col[1]) else f"{col[0]}_{col[1]}"
    if isinstance(col, tuple) else col
    for col in df.columns
]

df = df.drop(columns=["No"])

In [4]:
df.tail(10)

Company  \
98               Technosoft Consulting   
99                     Kompas Gramedia   
100                      Kredivo Group   
101                         Sportserve   
102  Mekari (PT. Mid Solusi Nusantara)   
103    ICS Compute Jakarta ICS Compute   
104                               Bjak   
105                            Gositus   
106                           NTT DATA   
107                             Shopee   

                                              Position  Location   \
98   Machine Learning Engineer – Intellegent Automo...  Indonesia   
99                 Machine Learning Engineer | Jakarta    Jakarta   
100                          Machine Learning Engineer    Jakarta   
101                            Indonesian AI Developer    Jakarta   
102                                        AI Engineer    Jakarta   
103                             Generative AI Engineer    Jakarta   
104                               AI Engineer (Remote)    Jakarta   
105                                        AI Engineer    Jakarta   
106                              Associate AI Engineer    Jakarta   
107                               AI Solution Engineer    Jakarta   

    WFH/ WFO/ Hybird length of work experience required (years)_MIN  \
98               NaN                                              5   
99               NaN                                              2   
100              NaN                                              2   
101              NaN                                            NaN   
102           Hybird                                            NaN   
103              NaN                                              2   
104              WFH                                              4   
105              NaN                                              2   
106           Hybird                                            NaN   
107              NaN                                              2   

    length of work experience required (years)_MAX Salary Access Date  \
98                                             NaN    NaN  2025-09-03   
99                                             NaN    NaN  2025-09-03   
100                                            NaN    NaN  2025-09-03   
101                                            NaN    NaN  2025-09-03   
102                                            NaN    NaN  2025-09-03   
103                                            NaN    NaN  2025-09-03   
104                                            NaN    NaN  2025-09-03   
105                                            NaN    NaN  2025-09-03   
106                                            NaN    NaN  2025-09-03   
107                                            NaN    NaN  2025-09-03   

    Upload By Company Enthusiast  ... Degree Information Platform_  Category  \
98          10 months        NaN  ...    NaN         NaN    Indeed       NaN   
99           2 months        NaN  ...    NaN         NaN    Indeed       NaN   
100         10 months        NaN  ...    NaN         NaN    Indeed       NaN   
101            1 week        NaN  ...    NaN         NaN    Indeed       NaN   
102           2 Weeks        NaN  ...    NaN         NaN    Indeed       NaN   
103           1 month        NaN  ...    NaN         NaN    Indeed       NaN   
104           1 month        NaN  ...    NaN         NaN    Indeed       NaN   
105            6 days        NaN  ...    NaN         NaN    Indeed       NaN   
106          2 months        NaN  ...    NaN         NaN    Indeed       NaN   
107           3 weeks        NaN  ...    NaN         NaN    Indeed       NaN   

    Degree Platform Quota   Category      Degree  \
98     NaN      NaN   NaN        NaN  Bachelor's   
99     NaN      NaN   1.0  Full-Time  Bachelor's   
100    NaN      NaN   NaN  Full-Time         NaN   
101    NaN      NaN   NaN        NaN  Bachelor's   
102    NaN      NaN   NaN        NaN  Bachelor's   
103    NaN      NaN   1.0  Fu

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 21 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         93 non-null     object        
 1   Position                                        93 non-null     object        
 2   Location                                        90 non-null     object        
 3   WFH/ WFO/ Hybird                                80 non-null     object        
 4   length of work experience required (years)_MIN  85 non-null     object        
 5   length of work experience required (years)_MAX  64 non-null     object        
 6   Salary                                          67 non-null     object        
 7   Access Date                                     92 non-null     datetime64[ns]
 8   Upload By Company                               86

In [6]:
df = df.drop(columns=["Platform"])
df = df.rename(columns={"Platform_": "Platform"})

## Merapihkan Kolom Duplikat

In [7]:
dupes = df.columns[df.columns.duplicated()].unique()
print("Kolom duplikat:", dupes)

Kolom duplikat: Index(['Category', 'Degree', 'Information'], dtype='object')


Terdapat kolom duplikat yang ada value nya jadi harus digabungkan dulu

In [8]:
for col in dupes:
    same_cols = [c for c in df.columns if c == col or c.startswith(col + ".")]
    print(f"{col}: {same_cols}")

Category: ['Category', 'Category', 'Category']
Degree: ['Degree', 'Degree', 'Degree']
Information: ['Information', 'Information']


Melihat apakah value ada disemua kolom duplikat atau hanya salah satu saja, kalau ada disemua kolom kita bisa drop salah satu tapi kalau hanya adalah di salah satu kita harus combine

In [9]:
# digunkaan untuk melihat index kolom
for i, col in enumerate(df.columns):
    print(i, col)

0 Company
1 Position
2 Location 
3 WFH/ WFO/ Hybird
4 length of work experience required (years)_MIN
5 length of work experience required (years)_MAX
6 Salary
7 Access Date
8 Upload By Company
9 Enthusiast
10 Category
11 Degree
12 Information
13 Platform
14 Category
15 Degree
16 Quota
17 Category
18 Degree
19 Information


In [10]:
df_degree = df[df.iloc[:,15].notna()]# index bisa diganti untuk melihat kolom duplikat lain
df_degree.head()


Company                              Position  \
39                      Kompas Gramedia  Machine Learning Engineer | Jakarta    
40  Nomura Research Institute Indonesia      Machine Learning Engineer (MLE)    
41                          ICS Compute               Generative AI Engineer    
42                              Gositus                          AI Engineer    
43       PT. Indocyber Global Teknologi                    AI Engineer Junior   

                     Location  WFH/ WFO/ Hybird  \
39     West Jakarta, Indonesia              WFO   
40    Jakarta Pusat, Indonesia              WFH   
41  Jakarta Selatan, Indonesia              WFO   
42  Central Jakarta, Indonesia              WFO   
43     West Jakarta, Indonesia              WFO   

   length of work experience required (years)_MIN  \
39                                              2   
40                                              3   
41                                              2   
42                                              2   
43                                              1   

   length of work experience required (years)_MAX  \
39                                              -   
40                                              -   
41                                              -   
42                                              -   
43                                              -   

                                       Salary Access Date Upload By Company  \
39                                          -  2025-08-02         - 2 month   
40  12,000,000.00 - 24,000,000.00 IDR / month  2025-08-02         - 5 month   
41                                          -  2025-08-02         - 1 month   
42                                          -  2025-08-02          - 5 days   
43                                          -  2025-08-02         - 2 month   

   Enthusiast Category Degree Information Platform     Category  \
39        NaN      NaN    NaN         NaN  Kalibrr    Full-Time   
40        NaN      NaN    NaN         NaN  Kalibrr    Full-Time   
41        NaN      NaN    NaN         NaN  Kalibrr    Full-Time   
42        NaN      NaN    NaN         NaN  Kalibrr    Full-Time   
43        NaN      NaN    NaN         NaN  Kalibrr  Contractual   

                    Degree  Quota Category Degree Information  
39  Bachelor's or Master's    NaN      NaN    NaN         NaN  
40                       -    NaN      NaN    NaN         NaN  
41  Bachelor's or Master's    NaN      NaN    NaN         NaN  
42              Bachelor's    NaN      NaN    NaN         NaN  
43              Bachelor's    NaN      NaN    NaN         NaN

Diantara setiap kolom duplikat hanya 1 kolom saja yang memiliki value (1 ada isi, lainnya kosong) sehingga kita combine jadi 1 kolom saja

In [11]:
dupes = df.columns[df.columns.duplicated()].unique()

for col in dupes:
    # Ambil semua kolom dengan nama sama
    same_cols = df.loc[:, df.columns == col]

    # Gabungkan ke satu kolom: ambil nilai pertama yang tidak null
    df[col] = same_cols.bfill(axis=1).iloc[:, 0]

    # Drop kolom duplikat lain (sisakan satu saja)

df = df.loc[:, ~df.columns.duplicated()]


C:\Users\User\AppData\Local\Temp\ipykernel_10404\1328924877.py:8: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = same_cols.bfill(axis=1).iloc[:, 0]
C:\Users\User\AppData\Local\Temp\ipykernel_10404\1328924877.py:8: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = same_cols.bfill(axis=1).iloc[:, 0]


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 15 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         93 non-null     object        
 1   Position                                        93 non-null     object        
 2   Location                                        90 non-null     object        
 3   WFH/ WFO/ Hybird                                80 non-null     object        
 4   length of work experience required (years)_MIN  85 non-null     object        
 5   length of work experience required (years)_MAX  64 non-null     object        
 6   Salary                                          67 non-null     object        
 7   Access Date                                     92 non-null     datetime64[ns]
 8   Upload By Company                               86

Kolom yang duplikat semuanya sudah tergabung

## Handling Null Value

In [13]:
len(df)

108

Next bisa dilihat kalau jumlah baris ada 108 sedangkan company hanya ada 93 (ini kolom yang mustahil kosong di data ini) artinya ada baris kosong yang sebaiknya dihapus saja

In [14]:
df[df["Company"].isnull()]

Company Position Location  WFH/ WFO/ Hybird  \
66     NaN      NaN       NaN              NaN   
67     NaN      NaN       NaN              NaN   
68     NaN      NaN       NaN              NaN   
69     NaN      NaN       NaN              NaN   
70     NaN      NaN       NaN              NaN   
71     NaN      NaN       NaN              NaN   
72     NaN      NaN       NaN              NaN   
73     NaN      NaN       NaN              NaN   
74     NaN      NaN       NaN              NaN   
75     NaN      NaN       NaN              NaN   
76     NaN      NaN       NaN              NaN   
77     NaN      NaN       NaN              NaN   
78     NaN      NaN       NaN              NaN   
79     NaN      NaN       NaN              NaN   
80     NaN      NaN       NaN              NaN   

   length of work experience required (years)_MIN  \
66                                            NaN   
67                                            NaN   
68                                            NaN   
69                                            NaN   
70                                            NaN   
71                                            NaN   
72                                            NaN   
73                                            NaN   
74                                            NaN   
75                                            NaN   
76                                            NaN   
77                                            NaN   
78                                            NaN   
79                                            NaN   
80                                            NaN   

   length of work experience required (years)_MAX Salary Access Date  \
66                                            NaN    NaN         NaT   
67                                            NaN    NaN         NaT   
68                                            NaN    NaN         NaT   
69                                            NaN    NaN         NaT   
70                                            NaN    NaN         NaT   
71                                            NaN    NaN         NaT   
72                                            NaN    NaN         NaT   
73                                            NaN    NaN         NaT   
74                                            NaN    NaN         NaT   
75                                            NaN    NaN         NaT   
76                                            NaN    NaN         NaT   
77                                            NaN    NaN         NaT   
78                                            NaN    NaN         NaT   
79                                            NaN    NaN         NaT   
80                                            NaN    NaN         NaT   

   Upload By Company Enthusiast Category Degree Information Platform  Quota  
66               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
67               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
68               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
69               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
70               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
71               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
72               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
73               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
74               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
75               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
76               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
77               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
78               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
79               NaN        NaN      NaN    NaN         NaN   dealls    NaN  
80               NaN        Na

In [15]:
df = df.dropna(how="all", subset=df.columns.difference(["Platform"]))

In [16]:
df[df["Company"].isnull()]

Empty DataFrame
Columns: [Company, Position, Location , WFH/ WFO/ Hybird, length of work experience required (years)_MIN, length of work experience required (years)_MAX, Salary, Access Date, Upload By Company, Enthusiast, Category, Degree, Information, Platform, Quota]
Index: []

## Menghapus baris yang duplikat

Dikarenakan data diambil dari berbagai platform tidak menutup kemungkinan ada lowongan pekerjaan dari company yang sama sehingga sebaiknya di drop saja

In [17]:
cols = ["Company", "Position"]

# buat versi normalisasi lowercase + strip spasi
normalized = df[cols].fillna("").apply(lambda c: c.str.lower().str.strip())

# kalau perlu, hapus multiple spaces di tengah jadi 1 spasi
normalized = normalized.apply(lambda c: c.str.replace(r"\s+", " ", regex=True))

# cek duplikat case-insensitive setelah normalisasi
mask = normalized.duplicated(keep=False)

# lihat hasil
df[mask]


Company  \
15                         Tap Growth ai   
19                            Sportserve   
21                                Shopee   
23                         Kredivo Group   
27                         Tap Growth ai   
39                       Kompas Gramedia   
40   Nomura Research Institute Indonesia   
42                               Gositus   
43        PT. Indocyber Global Teknologi   
83       PT. Indocyber Global Teknologi    
90   Nomura Research Institute Indonesia   
99                       Kompas Gramedia   
100                        Kredivo Group   
101                           Sportserve   
105                              Gositus   
107                               Shopee   

                                 Position  \
15           AI Engineer (Karawang -KIIC)   
19                Indonesian AI Developer   
21                   AI Solution Engineer   
23              Machine Learning Engineer   
27           AI Engineer (Karawang -KIIC)   
39   Machine Learning Engineer | Jakarta    
40       Machine Learning Engineer (MLE)    
42                           AI Engineer    
43                     AI Engineer Junior   
83                    AI Engineer Junior    
90        Machine Learning Engineer (MLE)   
99    Machine Learning Engineer | Jakarta   
100             Machine Learning Engineer   
101               Indonesian AI Developer   
105                           AI Engineer   
107                  AI Solution Engineer   

                                 Location  WFH/ WFO/ Hybird  \
15   Cikarang Barat, Jawa Barat, Indonesia              WFO   
19                               Indonesia           Hybird   
21                 Jakarta Raya, Indonesia              WFO   
23                        Area DKI Jakarta              WFO   
27   Cikarang Barat, Jawa Barat, Indonesia              WFO   
39                 West Jakarta, Indonesia              WFO   
40                Jakarta Pusat, Indonesia              WFH   
42              Central Jakarta, Indonesia              WFO   
43                 West Jakarta, Indonesia              WFO   
83                                 Jakarta              WFO   
90                                 Jakarta              WFH   
99                                 Jakarta              NaN   
100                                Jakarta              NaN   
101                                Jakarta              NaN   
105                                Jakarta              NaN   
107                                Jakarta              NaN   

    length of work experience required (years)_MIN  \
15                                               4   
19                                               -   
21                                               2   
23                                               2   
27                                               4   
39                                               2   
40                                               3   
42                                               2   
43                                               1   
83                                               1   
90                                               3   
99                                               2   
100                                              2   
101                                            NaN   
105                                              2   
107                                              2   

    length of work experience required (years)_MAX  \
15                                               -   
19                                               -   
21                                               -   
23                                               3   
27                                               -   
39                                               -   
40                                               -   
42                                               -   
43                                       

Bisa dilihat jika ada beberapa pekerjaan yang sama persis sehingga sebaiknya diambil salah satu aja

In [18]:
df = df.loc[~normalized.duplicated(keep="first")].reset_index(drop=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 15 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         85 non-null     object        
 1   Position                                        85 non-null     object        
 2   Location                                        82 non-null     object        
 3   WFH/ WFO/ Hybird                                77 non-null     object        
 4   length of work experience required (years)_MIN  78 non-null     object        
 5   length of work experience required (years)_MAX  63 non-null     object        
 6   Salary                                          65 non-null     object        
 7   Access Date                                     84 non-null     datetime64[ns]
 8   Upload By Company                               78 n

## Normalisasi value

In [19]:
df.tail(10)

Company  \
75                Samsung Electronics   
76                                 EY   
77                           Portcast   
78                              Bonza   
79                          Accenture   
80              Technosoft Consulting   
81  Mekari (PT. Mid Solusi Nusantara)   
82    ICS Compute Jakarta ICS Compute   
83                               Bjak   
84                           NTT DATA   

                                             Position  Location   \
75                                 Senior AI Engineer    Jakarta   
76  Machine Learning Engineering Consultant (Junio...    Jakarta   
77                   Senior Machine Learning Engineer  Singapore   
78                   Senior Machine Learning Engineer    Jakarta   
79                              Data & AI Engineering    Jakarta   
80  Machine Learning Engineer – Intellegent Automo...  Indonesia   
81                                        AI Engineer    Jakarta   
82                             Generative AI Engineer    Jakarta   
83                               AI Engineer (Remote)    Jakarta   
84                              Associate AI Engineer    Jakarta   

   WFH/ WFO/ Hybird length of work experience required (years)_MIN  \
75              WFO                                              5   
76              WFO                                              1   
77              WFH                                              5   
78              NaN                                            NaN   
79              NaN                                            NaN   
80              NaN                                              5   
81           Hybird                                            NaN   
82              NaN                                              2   
83              WFH                                              4   
84           Hybird                                            NaN   

   length of work experience required (years)_MAX Salary Access Date  \
75                                            NaN    NaN  2025-09-03   
76                                            NaN    NaN  2025-09-03   
77                                            NaN    NaN  2025-09-03   
78                                            NaN    NaN  2025-09-03   
79                                            NaN    NaN  2025-09-03   
80                                            NaN    NaN  2025-09-03   
81                                            NaN    NaN  2025-09-03   
82                                            NaN    NaN  2025-09-03   
83                                            NaN    NaN  2025-09-03   
84                                            NaN    NaN  2025-09-03   

   Upload By Company Enthusiast   Category            Degree  \
75           1 month        NaN  Full-Time  Ph.D (preferred)   
76           2 weeks        NaN  Full-Time        Bachelor's   
77          5 months        NaN  Full-Time        Bachelor's   
78         10 months        NaN  Full-Time        Bachelor's   
79               NaN        NaN  Full-Time               NaN   
80         10 months        NaN        NaN        Bachelor's   
81           2 Weeks        NaN        NaN        Bachelor's   
82           1 month        NaN  Full-Time        Bachelor's   
83           1 month        NaN  Full-Time        Bachelor's   
84          2 months        NaN  Full-Time        Bachelor's   

                                   Information Platform  Quota  
75                                         NaN   Indeed    NaN  
76                                         NaN   Indeed    NaN  
77                                         NaN   Indeed    NaN  
78                                         NaN   Indeed    NaN  
79  Biasanya memberikan tanggapan dalam 3 hari   Indeed    NaN  
80                                         NaN   Indeed    NaN  
81                                         NaN   Indeed    NaN  
82                                      

Terdapat beberapa kolom yang memiliki value tidak konsisten oleh karena itu normalisasi value perlu dilakukan untuk menyamaratakan 

### Kolom Location

In [20]:
df["Location "].unique()

array(['Asia Pasifik dan Jepang', 'Jakarta, Jakarta Raya, Indonesia',
       'Bekasi Selatan, Jawa Barat, Indonesia',
       'Jakarta, Jakarta Raya, Indonesia ',
       'Gedebage, Jawa Barat, Indonesia',
       'Kebayoran Lama, Jakarta Raya, Indonesia', 'Indonesia',
       'Asia Pasifik', 'DIY, DIY, Indonesia or Bali',
       'Bandung, Banten, Indonesia', 'Jakarta Raya, Indonesia',
       'DIY, DIY, Indonesia', 'Cikarang Barat, Jawa Barat, Indonesia',
       'Gambir, Jakarta Raya, Indonesia', 'Area DKI Jakarta',
       'Jakarta Raya, Indonesia ', 'Pondok Aren, Banten, Indonesia',
       'West Jakarta, Indonesia', 'Jakarta Pusat, Indonesia',
       'Jakarta Selatan, Indonesia', 'Central Jakarta, Indonesia', nan,
       'Jakarta, Jakarta Selatan, Indonesia', 'Jakarta',
       'Jakarta, Jakarta Timur, Indonesia', 'Bogor, Jawa Barat',
       'Surabaya', 'Yogyakarta', 'Jakarta, Jakarta Utara, Indonesia',
       'Tangerang', 'Jakarta Pusat', 'Singapore'], dtype=object)

Kolom location memiliki beberapa value yang seharusnya bisa jadi satu

In [21]:
location_map = {
    "Jakarta" : "Jakarta Raya, Indonesia",
    "Jakarta, Jakarta Raya, Indonesia": "Jakarta Raya, Indonesia",
    "Jakarta Raya, Indonesia ": "Jakarta Raya, Indonesia",
    "Area DKI Jakarta" : "Jakarta Raya, Indonesia",
    "Jakarta Pusat": "Jakarta Pusat, Indonesia",
    "Central Jakarta, Indonesia": "Jakarta Pusat, Indonesia",
    "West Jakarta, Indonesia": "Jakarta Barat, Indonesia",
    "Jakarta Selatan, Indonesia": "Jakarta Selatan, Indonesia",
    "Jakarta, Jakarta Selatan, Indonesia": "Jakarta Selatan, Indonesia",
    "Jakarta Pusat, Indonesia": "Jakarta Pusat, Indonesia",
    "Jakarta, Jakarta Utara, Indonesia": "Jakarta Utara, Indonesia",
    "Jakarta, Jakarta Timur, Indonesia": "Jakarta Timur, Indonesia",
    "Jakarta, Jakarta Raya, Indonesia ": "Jakarta Raya, Indonesia",
    "DIY, DIY, Indonesia": "Yogyakarta, Indonesia",
    "DIY, DIY, Indonesia or Bali": "Yogyakarta/Bali, Indonesia",
    "Yogyakarta" : "Yogyakarta, Indonesia",
    "Bandung, Banten, Indonesia": "Bandung, Jawa Barat, Indonesia",
}

df["Location "] = (
    df["Location "]
    .astype(str)
    .str.strip()
    .replace("nan", np.nan)
    .replace(location_map)
)

In [22]:
df["Location "].value_counts()

Location 
Jakarta Raya, Indonesia                    40
Jakarta Selatan, Indonesia                 10
Indonesia                                   6
Jakarta Pusat, Indonesia                    4
Asia Pasifik dan Jepang                     2
Jakarta Barat, Indonesia                    2
Asia Pasifik                                2
Yogyakarta, Indonesia                       2
Tangerang                                   1
Jakarta Utara, Indonesia                    1
Surabaya                                    1
Bogor, Jawa Barat                           1
Jakarta Timur, Indonesia                    1
Gambir, Jakarta Raya, Indonesia             1
Pondok Aren, Banten, Indonesia              1
Cikarang Barat, Jawa Barat, Indonesia       1
Bandung, Jawa Barat, Indonesia              1
Yogyakarta/Bali, Indonesia                  1
Kebayoran Lama, Jakarta Raya, Indonesia     1
Gedebage, Jawa Barat, Indonesia             1
Bekasi Selatan, Jawa Barat, Indonesia       1
Singapore               

### Kolom WFH/ WFO/ Hybird

In [23]:
df["WFH/ WFO/ Hybird"].value_counts()

WFH/ WFO/ Hybird
WFO       38
WFH       23
Hybird    10
Hybrid     5
-          1
Name: count, dtype: int64

terdapat inkonsistensi dari value hybird karena typo sehingga harus dirapihkan

In [24]:
df["WFH/ WFO/ Hybird"] = (df["WFH/ WFO/ Hybird"].replace({"Hybird": "Hybrid",}))

In [25]:
df["WFH/ WFO/ Hybird"].value_counts()

WFH/ WFO/ Hybird
WFO       38
WFH       23
Hybrid    15
-          1
Name: count, dtype: int64

### Kolom work experience

In [26]:
df["length of work experience required (years)_MIN"].unique()

array([3, 0, '<1', '-', 1, 4, 2, 5, nan, 'Fresh graduate'], dtype=object)

In [27]:
df["length of work experience required (years)_MAX"].unique()

array([5, '-', 3, 6, 4, nan, 2.0], dtype=object)

In [28]:
def work_exp(kolom):
    mapping = {
    "Fresh graduate": 0,
    "<1": 0,
    "-": np.nan
}
    df[kolom] = (df[kolom].replace(mapping).astype("Int64"))
    return df[kolom]

df["length of work experience required (years)_MIN"] = work_exp("length of work experience required (years)_MIN")
df["length of work experience required (years)_MAX"] = work_exp("length of work experience required (years)_MAX")

C:\Users\User\AppData\Local\Temp\ipykernel_10404\3864993543.py:7: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[kolom] = (df[kolom].replace(mapping).astype("Int64"))


In [29]:
df["length of work experience required (years)_MIN"].unique()

<IntegerArray>
[3, 0, <NA>, 1, 4, 2, 5]
Length: 7, dtype: Int64

In [30]:
df["length of work experience required (years)_MAX"].unique()

<IntegerArray>
[5, <NA>, 3, 6, 4, 2]
Length: 6, dtype: Int64

### Kolom Salary

In [31]:
df["Salary"].value_counts()

Salary
-                                            37
20.000.000 - 25.000.000                       1
Negotiable                                    1
6.000.000 – 7.500.000                         1
2025-09-03 00:00:00                           1
5.000.000 - 7.000.000                         1
3.000.000 - 4.500.000                         1
3.200.000 - 5.500.000                         1
3.000.000 - 5.000.000                         1
4.000.000 - 6.000.000                         1
4.000.000 - 8.000.000                         1
5.600.000 - 8.400.000                         1
5.400.000 - 8.000.000                         1
7.000.000 - 12.000.000                        1
7.000.000 - 9.000.000                         1
RP 1,5 - 2 M                                  1
5.000.000 - 9.000.000                         1
7.600.000 - 11.400.000                        1
8.000.000 - 9.000.000                         1
8.000.000 - 11.500.000                        1
20000000                         

Menormalisasi salary benar benar tricky karena value nya memiliki rentang, memiliki banyak noise, dan tidak konsisten (ada yang perhour)

In [32]:
df["salary_str"] = df["Salary"].astype(str).str.lower()

# 2. Hilangkan label umum
df["salary_str"] = (
    df["salary_str"]
    .str.replace("rp", "", regex=False)
    .str.replace("idr", "", regex=False)
    .str.replace("/bulan", "", regex=False)
    .str.replace("/ month", "", regex=False)
    .str.replace("per bulan", "", regex=False)
    .str.replace("per month", "", regex=False)
    # .str.replace("/ hour", "", regex=False)
    # .str.replace("per jam", "", regex=False)
    # .str.replace("per hour", "", regex=False)
    .str.replace("\u2005", "", regex=False)
    .str.replace("negotiable", "", regex=False)
)

# 3. Normalisasi whitespace (hapus spasi aneh)
df["salary_str"] = df["salary_str"].str.replace(r"\s+", " ", regex=True)

# 4. Standarisasi pemisah ribuan/desimal
df["salary_str"] = df["salary_str"].str.replace(",", ".", regex=False)
df["salary_str"] = df["salary_str"].str.replace(r"\.(?=\d{3})", "", regex=True)

# 5. Konversi k, m, ribu
df["salary_str"] = df["salary_str"].str.replace(
    r"(\d+(?:\.\d+)?)\s*k", 
    lambda m: str(float(m.group(1)) * 1_000), 
    regex=True
)

df["salary_str"] = df["salary_str"].str.replace(
    r"(\d+(?:\.\d+)?)\s*m", 
    lambda m: str(float(m.group(1)) * 1_000_000), 
    regex=True
)

df["salary_str"] = df["salary_str"].str.replace(
    r"(\d+(?:\.\d+)?)\s*ribu", 
    lambda m: str(float(m.group(1)) * 1_000), 
    regex=True
)

# 6. Trim spasi
df["salary_str"] = df["salary_str"].str.strip()


In [33]:
def normalize_range(val):
    if "-" not in val:
        return val
    parts = [p.strip() for p in val.split("-")]
    if len(parts) != 2:
        return val
    
    try:
        left = float(parts[0])
        right = float(parts[1])
    except ValueError:
        return val
    
    # Jika kanan jauh lebih besar (contoh 2000000), dan kiri masih kecil (contoh 1.5)
    if right >= 1_000_000 and left < 100:
        left *= 1_000_000
    elif right >= 1_000 and left < 100:
        left *= 1_000
    
    return f"{left} - {right}"

df["salary_str"] = df["salary_str"].apply(normalize_range)


In [34]:
# filter value yang ada tanggal nya
df.loc[df["salary_str"].str.match(r"\d{4}-\d{2}-\d{2}", na=False), "salary_str"] = ""

In [35]:
df["salary_str"].value_counts()

salary_str
-                              37
nan                            20
7600000.0 - 11400000.0          2
                                2
20000000.0 - 25000000.0         1
6000000 – 7500000               1
5000000.0 - 7000000.0           1
3000000.0 - 4500000.0           1
3200000.0 - 5500000.0           1
3000000.0 - 5000000.0           1
4000000.0 - 6000000.0           1
4000000.0 - 8000000.0           1
5600000.0 - 8400000.0           1
5400000.0 - 8000000.0           1
7000000.0 - 12000000.0          1
7000000.0 - 9000000.0           1
1500000.0 - 2000000.0           1
5000000.0 - 9000000.0           1
8000000.0 - 9000000.0           1
8000000.0 - 11500000.0          1
20000000                        1
12000000.0 - 24000000.0         1
310000.0 – 465000.0 per jam     1
3500000.0 - 4000000.0           1
231000.0 / hour                 1
165000.0 - 990000.0             1
132000.0 - 1070000.0 / hour     1
15750000.0 - 22500000.0         1
Name: count, dtype: int64

In [36]:
df["salary_str"].unique()
(df["salary_str"].unique())

array(['-', '1500000.0 - 2000000.0', '7600000.0 - 11400000.0',
       '132000.0 - 1070000.0 / hour', '165000.0 - 990000.0',
       '231000.0 / hour', '3500000.0 - 4000000.0',
       '310000.0 – 465000.0 per jam', '12000000.0 - 24000000.0', 'nan',
       '20000000', '8000000.0 - 11500000.0', '8000000.0 - 9000000.0',
       '5000000.0 - 9000000.0', '7000000.0 - 9000000.0',
       '20000000.0 - 25000000.0', '7000000.0 - 12000000.0',
       '5400000.0 - 8000000.0', '5600000.0 - 8400000.0',
       '4000000.0 - 8000000.0', '4000000.0 - 6000000.0',
       '3000000.0 - 5000000.0', '3200000.0 - 5500000.0',
       '3000000.0 - 4500000.0', '5000000.0 - 7000000.0', '',
       '6000000 – 7500000', '15750000.0 - 22500000.0'], dtype=object)

In [37]:
# Mengubah string ke number
import re
def parse_num_str(s):
    s = re.sub(r'[^\d\.,]', '', s)   # keep only digits, dot, comma
    if not s:
        return np.nan

    # heuristik pemisah ribuan/desimal
    if s.count('.') > 1 and s.count(',') == 0:
        # banyak titik → titik = pemisah ribuan
        s = s.replace('.', '')
    elif s.count(',') > 1 and s.count('.') == 0:
        # banyak koma → koma = pemisah ribuan
        s = s.replace(',', '')
    elif s.count('.') >= 1 and s.count(',') >= 1:
        # ada keduanya → lihat posisi terakhir
        if s.rfind('.') > s.rfind(','):
            # titik terakhir setelah koma -> titik dianggap decimal -> hapus koma
            s = s.replace(',', '')
        else:
            # koma terakhir setelah titik -> koma decimal, titik ribuan
            s = s.replace('.', '').replace(',', '.')
    else:
        # hanya satu jenis separator atau tidak ada -> ubah koma jadi titik (decimal)
        s = s.replace(',', '.')

    try:
        return float(s)
    except:
        return np.nan
    
df["numbers"] = df["salary_str"].apply(lambda x: re.findall(r"[\d\.,]+", x))  
df["numbers"] = df["numbers"].apply(lambda lst: [parse_num_str(n) for n in lst] if lst else [])

In [38]:
df["numbers"].value_counts()

numbers
[]                          59
[7600000.0, 11400000.0]      2
[20000000.0, 25000000.0]     1
[6000000.0, 7500000.0]       1
[5000000.0, 7000000.0]       1
[3000000.0, 4500000.0]       1
[3200000.0, 5500000.0]       1
[3000000.0, 5000000.0]       1
[4000000.0, 6000000.0]       1
[4000000.0, 8000000.0]       1
[5600000.0, 8400000.0]       1
[5400000.0, 8000000.0]       1
[7000000.0, 12000000.0]      1
[7000000.0, 9000000.0]       1
[1500000.0, 2000000.0]       1
[5000000.0, 9000000.0]       1
[8000000.0, 9000000.0]       1
[8000000.0, 11500000.0]      1
[20000000.0]                 1
[12000000.0, 24000000.0]     1
[310000.0, 465000.0]         1
[3500000.0, 4000000.0]       1
[231000.0]                   1
[165000.0, 990000.0]         1
[132000.0, 1070000.0]        1
[15750000.0, 22500000.0]     1
Name: count, dtype: int64

In [39]:
df["salary_min"] = df["numbers"].apply(lambda x: x[0] if len(x)>=1 else np.nan)
df["salary_max"] = df["numbers"].apply(lambda x: x[1] if len(x)>=2 else (x[0] if len(x)==1 else np.nan))

In [40]:
# konversi per-hour ke per-month (asumsi 8 jam × 22 hari kerja)
mask_hour = df["salary_str"].str.contains("hour|jam", case=False, na=False)
df.loc[mask_hour, ["salary_min","salary_max"]] = df.loc[mask_hour, ["salary_min","salary_max"]].apply(lambda col: col * 8 * 22)

In [41]:
df["salary_min"].unique()

array([      nan,  1500000.,  7600000., 23232000.,   165000., 40656000.,
        3500000., 54560000., 12000000., 20000000.,  8000000.,  5000000.,
        7000000.,  5400000.,  5600000.,  4000000.,  3000000.,  3200000.,
        6000000., 15750000.])

In [42]:
df["salary_max"].unique()

array([       nan, 2.0000e+06, 1.1400e+07, 1.8832e+08, 9.9000e+05,
       4.0656e+07, 4.0000e+06, 8.1840e+07, 2.4000e+07, 2.0000e+07,
       1.1500e+07, 9.0000e+06, 2.5000e+07, 1.2000e+07, 8.0000e+06,
       8.4000e+06, 6.0000e+06, 5.0000e+06, 5.5000e+06, 4.5000e+06,
       7.0000e+06, 7.5000e+06, 2.2500e+07])

In [43]:
df = df.drop(columns=["Salary", "salary_str", "numbers"])

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 16 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         85 non-null     object        
 1   Position                                        85 non-null     object        
 2   Location                                        82 non-null     object        
 3   WFH/ WFO/ Hybird                                77 non-null     object        
 4   length of work experience required (years)_MIN  61 non-null     Int64         
 5   length of work experience required (years)_MAX  23 non-null     Int64         
 6   Access Date                                     84 non-null     datetime64[ns]
 7   Upload By Company                               78 non-null     object        
 8   Enthusiast                                      39 n

In [45]:
pd.set_option("display.float_format", "{:,.0f}".format)

In [46]:
df.head()

Company                                           Position  \
0             Braintrust                       Prompt Engineer / AI Engineer   
1   VIDA Digital Identity  Deep Learning / AI Scientist - Liveness Detect...   
2             RH USA, Inc                               AI Prompter Engineer   
3                Bibit.id                                     AI/ML Engineer   
4  CAD IT Gruppo Cedacri                              Computer Vision Intern   

                               Location  WFH/ WFO/ Hybird  \
0                Asia Pasifik dan Jepang              WFH   
1                Jakarta Raya, Indonesia                -   
2  Bekasi Selatan, Jawa Barat, Indonesia              WFO   
3                Jakarta Raya, Indonesia           Hybrid   
4        Gedebage, Jawa Barat, Indonesia              WFO   

   length of work experience required (years)_MIN  \
0                                               3   
1                                               3   
2                                               0   
3                                               0   
4                                               0   

   length of work experience required (years)_MAX Access Date  \
0                                               5  2025-08-30   
1                                            <NA>  2025-08-30   
2                                            <NA>  2025-08-30   
3                                            <NA>  2025-08-30   
4                                            <NA>  2025-08-30   

        Upload By Company Enthusiast   Category                  Degree  \
0               - 2 weeks      >100   Full-time                       -   
1  30-Agu-2025 (reupload)       >100  Full-time         master’s or PhD   
2                - 6 days         47  Full-time              Bachelor's   
3               - 1 Month       >100  Full-time  Bachelor's or Master's   
4     2025-08-30 00:00:00         34     Intern                 Diploma   

                         Information  Platform  Quota  salary_min  salary_max  
0  Respons dikelola di luar LinkedIn  Linkedin    NaN         NaN         NaN  
1  Respons dikelola di luar LinkedIn  Linkedin    NaN         NaN         NaN  
2  Respons dikelola di luar LinkedIn  Linkedin    NaN         NaN         NaN  
3                                  -  Linkedin    NaN         NaN         NaN  
4  Respons dikelola di luar LinkedIn  Linkedin    NaN   1,500,000   2,000,000

### Kolom Upload by Company

In [47]:
df["Upload By Company"].value_counts()

Upload By Company
- 1 Month                 10
- 2 weeks                  6
- 3 Weeks                  4
10 months                  3
2 months                   3
- 2 Days                   3
- 4 Days                   3
- 1 Weeks                  3
1 month                    3
- 1 Days                   2
- 5 days                   2
1 Month ago                2
11 days ago                2
27 days ago                2
3 weeks                    2
- 2 month                  2
5 months                   1
8 month ago                1
4 weeks                    1
1 months                   1
3 months                   1
2 weeks                    1
1 weeks                    1
9 days ago                 1
6 months                   1
12 days ago                1
20 days ago                1
8 days ago                 1
16 days ago                1
30-Agu-2025 (reupload)     1
! year ago                 1
- 1 month                  1
- 5 month                  1
this day                 

In [48]:
len(df["Upload By Company"].unique())

43

Kolom upload by company memiilki value yang beragam dan mostly dalam bentuk "months, weeks, dan day". Ini perlu dibersihkan dan sebaiknya diubah kedalam bentuk tanggal agar mudah dianalisis nantinya, dan karena ada kolom "access date" itu bisa menjadi acuan menentukan tanggal nya meskipun tidak 100% akurat

In [49]:
from datetime import timedelta

def clean_upload(val, access_date):
    if pd.isna(val):
        return np.nan
    
    # kalau sudah datetime object
    if isinstance(val, pd.Timestamp):
        return val
    
    val = str(val).lower().strip()
    
    val = re.split(r"\(", val)[0].strip()
    
    # special cases
    if val in ["this day", "today"]:
        return access_date
    if "! year ago" in val:
        return access_date - timedelta(days=365)
    
    # regex untuk days, weeks, months
    match = re.search(r"(\d+)\s*(day|days|week|weeks|month|months|moth|moths)", val)
    if match:
        num = int(match.group(1))
        unit = match.group(2)
        if "day" in unit:
            return access_date - timedelta(days=num)
        elif "week" in unit:
            return access_date - timedelta(weeks=num)
        elif "month" in unit or "moth" in unit:  # termasuk typo
            return access_date - timedelta(days=num*30)
    
    # coba parse langsung tanggal (misal 30-Agu-2025)
    parsed = pd.to_datetime(val, errors="coerce", dayfirst=True)
    if not pd.isna(parsed):
        return parsed
    
    return np.nan

# terapkan ke dataframe
df["Upload By Company"] = df.apply(
    lambda row: clean_upload(row["Upload By Company"], row["Access Date"]), axis=1
)

C:\Users\User\AppData\Local\Temp\ipykernel_10404\4013956673.py:34: UserWarning: Parsing dates in %Y-%m-%d %H:%M:%S format when dayfirst=True was specified. Pass `dayfirst=False` or specify a format to silence this warning.
  parsed = pd.to_datetime(val, errors="coerce", dayfirst=True)


In [50]:
df["Upload By Company"].value_counts()

Upload By Company
2025-08-02    5
2025-08-23    5
2025-08-04    5
2025-08-28    4
2025-08-26    4
2025-08-17    4
2025-08-16    3
2024-11-07    3
2025-08-11    3
2025-07-31    2
2025-08-07    2
2025-07-04    2
2025-06-03    2
2025-08-03    2
2025-08-12    2
2025-08-30    2
2025-08-01    2
2025-08-29    2
2025-08-20    2
2025-06-04    1
2025-08-25    1
2025-03-06    1
2025-08-05    1
2025-04-06    1
2025-01-06    1
2025-08-22    1
2025-08-13    1
2025-03-05    1
2025-08-18    1
2025-08-14    1
2024-09-02    1
2025-07-28    1
2025-07-03    1
2025-08-24    1
2025-08-31    1
2025-09-02    1
2025-08-27    1
2025-07-12    1
2025-05-02    1
2025-07-05    1
Name: count, dtype: int64

In [51]:
df["Upload By Company"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 85 entries, 0 to 84
Series name: Upload By Company
Non-Null Count  Dtype         
--------------  -----         
77 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 812.0 bytes


### Kolom Enthusiast

membersihkan value yang ada ">" karena ini masih berbentuk object

In [52]:
df["Enthusiast"].value_counts()

Enthusiast
 >100     21
 47        2
 >100      1
 59        1
 19        1
 22        1
 17        1
 18        1
 64        1
 77        1
 63        1
 38        1
 79        1
 74        1
 80        1
 65        1
 34        1
20         1
Name: count, dtype: int64

In [53]:
def clean_enthusiast(val):
    val = str(val).strip()
    if val.startswith(">"):
        return int(val[1:])  # ambil angka setelah ">"
    elif val.isdigit():
        return int(val)
    else:
        return np.nan  # kalau ada yang aneh
        

df["Enthusiast"] = df["Enthusiast"].apply(clean_enthusiast)

In [54]:
df["Enthusiast"].head()

0   100
1   100
2    47
3   100
4    34
Name: Enthusiast, dtype: float64

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 16 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         85 non-null     object        
 1   Position                                        85 non-null     object        
 2   Location                                        82 non-null     object        
 3   WFH/ WFO/ Hybird                                77 non-null     object        
 4   length of work experience required (years)_MIN  61 non-null     Int64         
 5   length of work experience required (years)_MAX  23 non-null     Int64         
 6   Access Date                                     84 non-null     datetime64[ns]
 7   Upload By Company                               77 non-null     datetime64[ns]
 8   Enthusiast                                      38 n

### Kolom category

kolom ini juga masih tak konsisten ada yang lowercase, menggunakan -, dsb

In [56]:
df["Category"].value_counts()

Category
Full-time             44
Full-Time             19
Contract               6
Full Time              3
Intern                 2
part time              2
Freelance              2
Contractual            1
Full-Time/Contract     1
Part Time              1
Name: count, dtype: int64

In [57]:
def clean_category(val):
    if pd.isna(val):
        return np.nan
    val = val.lower().strip()
    
    # hapus tanda baca umum
    val = re.sub(r"[-_/]", " ", val)  # ganti -, _, / jadi spasi
    val = re.sub(r"\s+", " ", val)    # rapikan spasi
    
    # mapping manual
    mapping = {
        "full time": "Full-time",
        "part time": "Part-time",
        "contract": "Contract",
        "contractual": "Contract",
        "freelance": "Freelance",
        "intern": "Intern",
    }
    
    return mapping.get(val, val.title())  # fallback: capitalisasi normal

df["Category"] = df["Category"].apply(clean_category)

In [58]:
df["Category"].value_counts()

Category
Full-time             66
Contract               7
Part-time              3
Intern                 2
Freelance              2
Full Time Contract     1
Name: count, dtype: int64

### Kolom Degree

In [59]:
df["Degree"].value_counts()

Degree
Bachelor's                         46
-                                  13
Bachelor's or Master's             10
Diploma                             5
master’s or PhD                     4
Master’s degree in STEM             1
Bachelor, Diploma 3, Diploma 4      1
Senior High School                  1
Ph.D (preferred)                    1
Name: count, dtype: int64

Kolom degree sebenarnya sudah unique semua tapi sebaiknya dirapikan 

In [60]:
def clean_degree(val):
    if pd.isna(val):
        return "Not Specified"
    
    val = val.lower().strip()
    
    if val in ["-", "none"]:
        return "Not Specified"
    if "high school" in val:
        return "High School"
    if "diploma" in val:
        return "Diploma"
    if "bachelor" in val:
        return "Bachelor"
    if "master" in val and "ph" not in val:
        return "Master"
    if "phd" in val or "ph.d" in val:
        return "PhD"
    if "bachelor's or master's" in val:
        return "Bachelor or Master"
    if "master’s or phd" in val:
        return "Master or PhD"
    
    return val.title()

df["Degree"] = df["Degree"].apply(clean_degree)

In [61]:
df["Degree"].value_counts()

Degree
Bachelor         56
Not Specified    16
Diploma           6
PhD               5
Master            1
High School       1
Name: count, dtype: int64

In [62]:
df.head(10)

Company  \
0                             Braintrust    
1                   VIDA Digital Identity   
2                             RH USA, Inc   
3                                Bibit.id   
4                  CAD IT Gruppo Cedacri    
5                Yoona Digital Indonesia    
6            Invisible Expert Marketplace   
7  SoftAge Information-Technology-Limited   
8                              Supercoder   
9                       ZettaByte Pte Ltd   

                                            Position  \
0                      Prompt Engineer / AI Engineer   
1  Deep Learning / AI Scientist - Liveness Detect...   
2                               AI Prompter Engineer   
3                                     AI/ML Engineer   
4                             Computer Vision Intern   
5  AI Chatbot Developer / Conversational AI Engineer   
6                Engineering Specialist – AI Trainer   
7        AI Agent Trainer (Remote – Flexible Hours)    
8                       AI Engineer (CAD Automation)   
9                                        AI Engineer   

                                 Location  WFH/ WFO/ Hybird  \
0                  Asia Pasifik dan Jepang              WFH   
1                  Jakarta Raya, Indonesia                -   
2    Bekasi Selatan, Jawa Barat, Indonesia              WFO   
3                  Jakarta Raya, Indonesia           Hybrid   
4          Gedebage, Jawa Barat, Indonesia              WFO   
5  Kebayoran Lama, Jakarta Raya, Indonesia              WFO   
6                                Indonesia              WFH   
7                                Indonesia              WFH   
8                             Asia Pasifik              WFH   
9               Yogyakarta/Bali, Indonesia              WFO   

   length of work experience required (years)_MIN  \
0                                               3   
1                                               3   
2                                               0   
3                                               0   
4                                               0   
5                                               0   
6                                            <NA>   
7                                            <NA>   
8                                            <NA>   
9                                               1   

   length of work experience required (years)_MAX Access Date  \
0                                               5  2025-08-30   
1                                            <NA>  2025-08-30   
2                                            <NA>  2025-08-30   
3                                            <NA>  2025-08-30   
4                                            <NA>  2025-08-30   
5                                            <NA>  2025-08-30   
6                                            <NA>  2025-08-30   
7                                            <NA>  2025-08-30   
8                                            <NA>  2025-08-30   
9                                            <NA>  2025-08-30   

  Upload By Company  Enthusiast   Category         Degree  \
0        2025-08-16         100  Full-time  Not Specified   
1               NaT         100  Full-time            PhD   
2        2025-08-24          47  Full-time       Bachelor   
3        2025-07-31         100  Full-time       Bachelor   
4        2025-08-30          34     Intern        Diploma   
5        2025-08-16         100  Full-time  Not Specified   
6        2025-08-16         100   Contract            PhD   
7        2025-08-23         100  Part-time  Not Specified   
8        2025-08-23         100  Full-time  Not Specified   
9        2025-05-02         100  Full-time  Not Specified   

                         Information  Platform  Quota  salary_min  salary_max  
0  Respons dikelola di luar LinkedIn  Linkedin    NaN         NaN         NaN  
1  Respons dikelola di luar LinkedIn  Linkedin    NaN         NaN         NaN  
2  Respons dike

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 16 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         85 non-null     object        
 1   Position                                        85 non-null     object        
 2   Location                                        82 non-null     object        
 3   WFH/ WFO/ Hybird                                77 non-null     object        
 4   length of work experience required (years)_MIN  61 non-null     Int64         
 5   length of work experience required (years)_MAX  23 non-null     Int64         
 6   Access Date                                     84 non-null     datetime64[ns]
 7   Upload By Company                               77 non-null     datetime64[ns]
 8   Enthusiast                                      38 n

### Mengubah Value Nan atau "-" di kolom tertentu 

In [64]:
# misalnya kolom yang ingin di-handle
cols_to_fix = ["Location ", "WFH/ WFO/ Hybird", "Category", "Degree"]  

for col in cols_to_fix:
    df[col] = df[col].replace("-", "Not Specified")   # ganti tanda -
    df[col] = df[col].fillna("Not Specified")         # ganti NaN

df["Enthusiast"] = df["Enthusiast"].replace("-", "No information").fillna("No information")
df["Information"] = df["Information"].replace("-", "No additional information").fillna("No additional information")


## Additional Preprocessing

#### Mengganti nama dan menghapus spasi berlebih pada judul kolom

In [65]:
print(df.columns.tolist())
df.columns = df.columns.str.strip()  
print(df.columns.tolist())

['Company', 'Position', 'Location ', 'WFH/ WFO/ Hybird', 'length of work experience required (years)_MIN', 'length of work experience required (years)_MAX', 'Access Date', 'Upload By Company', 'Enthusiast', 'Category', 'Degree', 'Information', 'Platform', 'Quota', 'salary_min', 'salary_max']
['Company', 'Position', 'Location', 'WFH/ WFO/ Hybird', 'length of work experience required (years)_MIN', 'length of work experience required (years)_MAX', 'Access Date', 'Upload By Company', 'Enthusiast', 'Category', 'Degree', 'Information', 'Platform', 'Quota', 'salary_min', 'salary_max']


In [66]:
df = df.rename(columns={
    "WFH/ WFO/ Hybird": "Work Arrangement",
    "Upload By Company": "Upload_Date"
})

#### Menghapus kolom quota karena mostly tidak bernilai

In [67]:
df = df.drop(columns=["Quota"])

In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 15 columns):
 #   Column                                          Non-Null Count  Dtype         
---  ------                                          --------------  -----         
 0   Company                                         85 non-null     object        
 1   Position                                        85 non-null     object        
 2   Location                                        85 non-null     object        
 3   Work Arrangement                                85 non-null     object        
 4   length of work experience required (years)_MIN  61 non-null     Int64         
 5   length of work experience required (years)_MAX  23 non-null     Int64         
 6   Access Date                                     84 non-null     datetime64[ns]
 7   Upload_Date                                     77 non-null     datetime64[ns]
 8   Enthusiast                                      85 n

In [69]:
df.head(20)

Company  \
0                              Braintrust    
1                    VIDA Digital Identity   
2                              RH USA, Inc   
3                                 Bibit.id   
4                   CAD IT Gruppo Cedacri    
5                 Yoona Digital Indonesia    
6             Invisible Expert Marketplace   
7   SoftAge Information-Technology-Limited   
8                               Supercoder   
9                        ZettaByte Pte Ltd   
10                                  Mekari   
11            PT DAYA CIPTA MANDIRI SOLUSI   
12                               GMI cloud   
13                               RWS Group   
14                               BitHealth   
15                           Tap Growth ai   
16                                   Pintu   
17                                    BJAK   
18                                Synapsis   
19                              Sportserve   

                                             Position  \
0                       Prompt Engineer / AI Engineer   
1   Deep Learning / AI Scientist - Liveness Detect...   
2                                AI Prompter Engineer   
3                                      AI/ML Engineer   
4                              Computer Vision Intern   
5   AI Chatbot Developer / Conversational AI Engineer   
6                 Engineering Specialist – AI Trainer   
7         AI Agent Trainer (Remote – Flexible Hours)    
8                        AI Engineer (CAD Automation)   
9                                         AI Engineer   
10                                        AI Engineer   
11      Medical Artificial Intelligence (AI) Engineer   
12   Global AI Cloud Inference System Engineer Intern   
13  AI Data Annotators – Indonesian | STEM Backgro...   
14                             Associate AI Engineer    
15                       AI Engineer (Karawang -KIIC)   
16                                        AI Engineer   
17     Machine Learning Engineer (Diffusion / Vision)   
18                                    Jr. AI Engineer   
19                            Indonesian AI Developer   

                                   Location Work Arrangement  \
0                   Asia Pasifik dan Jepang              WFH   
1                   Jakarta Raya, Indonesia    Not Specified   
2     Bekasi Selatan, Jawa Barat, Indonesia              WFO   
3                   Jakarta Raya, Indonesia           Hybrid   
4           Gedebage, Jawa Barat, Indonesia              WFO   
5   Kebayoran Lama, Jakarta Raya, Indonesia              WFO   
6                                 Indonesia              WFH   
7                                 Indonesia              WFH   
8                              Asia Pasifik              WFH   
9                Yogyakarta/Bali, Indonesia              WFO   
10                  Jakarta Raya, Indonesia           Hybrid   
11           Bandung, Jawa Barat, Indonesia              WFO   
12                  Asia Pasifik dan Jepang              WFH   
13                  Jakarta Raya, Indonesia              WFH   
14                    Yogyakarta, Indonesia              WFO   
15    Cikarang Barat, Jawa Barat, Indonesia              WFO   
16                                Indonesia              WFO   
17                  Jakarta Raya, Indonesia              WFH   
18          Gambir, Jakarta Raya, Indonesia              WFO   
19                                Indonesia           Hybrid   

    length of work experience required (years)_MIN  \
0                                                3   
1                                                3   
2                                                0   
3                                                0   
4                                                0   
5                                                0   
6                                             <NA>   
7                                             <NA>   
8                                      

### menyimpan data yang telah bersih

In [70]:
df.to_excel("../dataset/ML Engineer Jobs Cleaned.xlsx", index=False, engine="openpyxl")